networkx/algorithms/d_separation.py

```diff
@@ -0,0 +1,122 @@
+"""
+Algorithm for testing d-separation in DAGs.
+
+*d-separation* is a test for conditional independence in probability
+distributions that can be factorized using DAGs.  It is a purely
+graphical test that uses the underlying graph and makes no reference
+to the actual distribution parameters.  For a detailed treatment see
+[1]_, [2]_.
+"""
+
+from collections import deque
+from typing import Hashable, AbstractSet
+
+import networkx as nx
+from networkx.utils import not_implemented_for, UnionFind
+
+__all__ = ["d_separated"]
+
+
+@not_implemented_for("undirected")
+def d_separated(G: nx.DiGraph, x: AbstractSet[Hashable],
+                y: AbstractSet[Hashable], z: AbstractSet[Hashable]) -> bool:
+    """
+    Test whether two sets of nodes are d-separated in a DAG by a third
+    set of conditioning nodes.
+
+    Parameters
+    ----------
+    G : graph
+        A NetworkX DAG.
+
+    x : set
+        First set of nodes in `G`.
+
+    y : set
+        Second set of nodes in `G`.
+
+    z : set
+        Set of conditioning nodes in `G`.
```

Comment:
(copy/pasted) please change backticks.

```suggestion
        Set of conditioning nodes in ``G``. Can be empty set.
```

Reply:
The implementation allows all three sets to be empty. If `x` or `y` is empty the result is trivially True, since there are no conditional independence tests to be made. 

So, I feel that the elaboration (that the set can be empty) should be made for all or none. Your thoughts ?

Reply:
That is true. The trivial result is that "nothing is d-separated from nothing by nothing", which is quite hilarious now that I think about it.

That said, thinking from the perspective of how causal inference tasks usually go, I _think_ it is true that most of the time `x` and `y` will not be empty sets, while `z` might or might not be. The intent here is to nudge in the correct pedagogical direction, especially in a situation where we don't want to be _too_ strict with the restrictions. That is why I placed the `can be empty set` elaboration only at the docstring part for `z`.